pkg/storing/nodes/nodes_json.go

```diff
@@ -0,0 +1,266 @@
+	for i, node := range n {
+		if node.URL == updated.URL {
+			n[i] = updated
+			return true
+		}
+	}
+	return false
+}
+
+func appendIfNew(ns nodes, url string) (nodes, bool) {
+	for _, node := range ns {
+		if node.URL == url {
+			return ns, false
+		}
+	}
+	newNode := entities.Node{URL: url, Enabled: true}
+	return append(ns, newNode), true
+}
+
+func deleteIfFound(ns nodes, url string) (nodes, bool) {
+	for i, node := range ns {
+		if node.URL == url {
+			ns[i] = ns[len(ns)-1] // replace found element with the last element
+			return ns[:len(ns)-1], true
+		}
+	}
+	return ns, false
+}
+
+type dbStruct struct {
+	CommonNodes   nodes `json:"common_nodes"`
+	SpecificNodes nodes `json:"specific_nodes"`
+}
+
+func (db dbStruct) Nodes(specific bool) nodes {
+	if specific {
+		return db.SpecificNodes
+	}
+	return db.CommonNodes
+}
+
+func (db dbStruct) SyncToFile(path string) error {
+	data, err := json.MarshalIndent(db, "", " ")
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal nodes db as json")
+	}
+	if err := os.WriteFile(path, data, 0600); err != nil {
+		return errors.Wrapf(err, "failed to write nodes db data to file '%s'", path)
+	}
+	return nil
+}
+
+type JSONStorage struct {
+	mu     *sync.RWMutex
+	db     *dbStruct
+	dbFile string
+	zap    *zap.Logger
+}
+
+func NewJSONStorage(path string, nodes []string, logger *zap.Logger) (*JSONStorage, error) {
+	path = filepath.Clean(path)
+	if err := createDBFileIfNotExist(path); err != nil {
+		return nil, errors.Wrapf(err, "failed to create and init nodes db file '%s'", path)
+	}
+
+	data, err := os.ReadFile(path) // now we have some guarantees that file exists
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read nodes file")
+	}
+	db := new(dbStruct)
+	if err := json.Unmarshal(data, db); err != nil {
+		return nil, errors.Wrapf(err, "failed to parse data as json from file '%s'", path)
+	}
+	s := &JSONStorage{mu: new(sync.RWMutex), db: db, dbFile: path, zap: logger}
+	if err := s.populate(nodes); err != nil {
+		return nil, errors.Wrapf(err, "failed to populate nodes storage")
+	}
+	return s, nil
+}
+
+func createDBFileIfNotExist(path string) error {
+	switch _, err := os.Stat(path); {
+	case err == nil:
+		return nil
+	case errors.Is(err, os.ErrNotExist):
+		var empty dbStruct
+		data, err := json.Marshal(empty)
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal empty db struct to json")
+		}
+		return os.WriteFile(path, data, 0600)
+	default:
+		return errors.Wrap(err, "failed to get stat for file")
+	}
+}
+
+func (s *JSONStorage) Close() error {
+	return nil
+}
+
+func (s *JSONStorage) Nodes(specific bool) ([]entities.Node, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.db.Nodes(specific), nil
+}
+
+func (s *JSONStorage) EnabledNodes() ([]entities.Node, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.db.CommonNodes.queryNodes(func(n entities.Node) bool { return n.Enabled }, 0), nil
```

Comment:
We probably don't need `Enabled` and `Limit` fields anymore, we never really used it. So instead of `EnabledNodes` and `EnabledSpecificNodes` I would just use `Nodes`

Reply:
Agree. Though, on my mind, it should be done in a new PR because this change will cause some significant changes in other packages and tests. Please, create issue and assign it on me